File: tests/Aiko.Core.Tests/SnapshotFileTests.cs
namespace Aiko.Core.Tests;

public class SnapshotFileTests
{
    private static readonly DateTimeOffset Now = new(2026, 9, 12, 12, 0, 0, TimeSpan.Zero);

    [Fact]
    public void What_the_bridge_writes_the_tray_reads_back()
    {
        var snapshot = new LimitSnapshot(
            "Personal",
            LimitSource.StatusLine,
            Now,
            [new LimitWindow(LimitKind.FiveHour, 34, Now.AddHours(3))]);

        var again = SnapshotFile.FromJson(SnapshotFile.ToJson(snapshot));

        Assert.Equal("Personal", again.Environment);
        Assert.Equal(LimitSource.StatusLine, again.Source);
        Assert.Equal(Now, again.ReceivedAt);
        Assert.Equal(34, again.Windows.Single().Percent);
        Assert.Equal(Now.AddHours(3), again.Windows.Single().ResetsAt);
    }

    [Fact]
    public void The_model_limit_survives_the_round_trip()
    {
        var snapshot = new LimitSnapshot(
            "Personal",
            LimitSource.DirectMode,
            Now,
            [new LimitWindow(LimitKind.FiveHour, 10, Now.AddHours(1))])
        {
            Model = new ModelLimit("Fable", 3, Now.AddDays(4)),
        };

        var again = SnapshotFile.FromJson(SnapshotFile.ToJson(snapshot));

        Assert.Equal("Fable", again.Model!.Value.ModelName);
        Assert.Equal(3, again.Model.Value.Percent);
        Assert.Equal(LimitSource.DirectMode, again.Source);
    }

    [Fact]
    public void The_file_holds_numbers_only_and_no_traces_of_the_session()
    {
        var snapshot = new LimitSnapshot(
            "Personal",
            LimitSource.StatusLine,
            Now,
            [new LimitWindow(LimitKind.FiveHour, 34, Now.AddHours(3))]);

        var json = SnapshotFile.ToJson(snapshot);

        Assert.DoesNotContain("session", json, StringComparison.OrdinalIgnoreCase);
        Assert.DoesNotContain("transcript", json, StringComparison.OrdinalIgnoreCase);
        Assert.DoesNotContain("cwd", json, StringComparison.OrdinalIgnoreCase);
        Assert.DoesNotContain("token", json, StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public void Enums_are_written_as_words_so_the_file_stays_readable()
    {
        var json = SnapshotFile.ToJson(new LimitSnapshot(
            "Work",
            LimitSource.DirectMode,
            Now,
            [new LimitWindow(LimitKind.SevenDay, 61, Now.AddDays(4))]));

        Assert.Contains("\"DirectMode\"", json);
        Assert.Contains("\"SevenDay\"", json);
    }

    [Theory]
    [InlineData("")]
    [InlineData("half written {")]
    [InlineData("{}")]
    [InlineData("{ \"environment\": \"Personal\", \"windows\": [] }")]
    public void A_file_being_written_right_now_reads_as_nothing_new(string json)
    {
        var snapshot = SnapshotFile.FromJson(json, "Personal");

        Assert.False(snapshot.HasData);
        Assert.Equal("Personal", snapshot.Environment);
    }
}

File: mac/Tests/AikoKitTests/SnapshotFileTests.swift
import Foundation
import Testing

@testable import AikoKit

struct SnapshotFileTests {
    static let now = utc(2026, 9, 12, 12, 0, 0)

    @Test
    func whatTheBridgeWritesTheTrayReadsBack() {
        let snapshot = LimitSnapshot(
            environment: "Personal",
            source: .statusLine,
            receivedAt: Self.now,
            windows: [LimitWindow(kind: .fiveHour, percent: 34, resetsAt: Self.now.adding(hours: 3))])

        let again = SnapshotFile.fromJson(SnapshotFile.toJson(snapshot))

        #expect(again.environment == "Personal")
        #expect(again.source == .statusLine)
        #expect(again.receivedAt == Self.now)
        #expect(again.windows.first?.percent == 34)
        #expect(again.windows.first?.resetsAt == Self.now.adding(hours: 3))
    }

    @Test
    func theModelLimitSurvivesTheRoundTrip() {
        let snapshot = LimitSnapshot(
            environment: "Personal",
            source: .directMode,
            receivedAt: Self.now,
            windows: [LimitWindow(kind: .fiveHour, percent: 10, resetsAt: Self.now.adding(hours: 1))],
            model: ModelLimit(modelName: "Fable", percent: 3, resetsAt: Self.now.adding(days: 4)))

        let again = SnapshotFile.fromJson(SnapshotFile.toJson(snapshot))

        #expect(again.model?.modelName == "Fable")
        #expect(again.model?.percent == 3)
        #expect(again.source == .directMode)
    }

    @Test
    func theFileHoldsNumbersOnlyAndNoTracesOfTheSession() {
        let snapshot = LimitSnapshot(
            environment: "Personal",
            source: .statusLine,
            receivedAt: Self.now,
            windows: [LimitWindow(kind: .fiveHour, percent: 34, resetsAt: Self.now.adding(hours: 3))])

        let json = SnapshotFile.toJson(snapshot).lowercased()

        #expect(!json.contains("session"))
        #expect(!json.contains("transcript"))
        #expect(!json.contains("cwd"))
        #expect(!json.contains("token"))
    }

    @Test
    func enumsAreWrittenAsWordsSoTheFileStaysReadable() {
        let json = SnapshotFile.toJson(
            LimitSnapshot(
                environment: "Work",
                source: .directMode,
                receivedAt: Self.now,
                windows: [LimitWindow(kind: .sevenDay, percent: 61, resetsAt: Self.now.adding(days: 4))]))

        #expect(json.contains("\"DirectMode\""))
        #expect(json.contains("\"SevenDay\""))
    }

    @Test(arguments: ["", "half written {", "{}", #"{ "environment": "Personal", "windows": [] }"#])
    func aFileBeingWrittenRightNowReadsAsNothingNew(json: String) {
        let snapshot = SnapshotFile.fromJson(json, fallbackEnvironment: "Personal")

        #expect(!snapshot.hasData)
        #expect(snapshot.environment == "Personal")
    }

    @Test
    func aTimeIsWrittenTheWayTheWindowsCoreWritesIt() {
        // System.Text.Json writes a DateTimeOffset with the seconds and the offset, and drops a
        // fraction that is zero. Its strict encoder also escapes the plus of the offset, so both
        // cores put the same bytes in the file.
        let json = SnapshotFile.toJson(
            LimitSnapshot(
                environment: "Work",
                source: .statusLine,
                receivedAt: Self.now,
                windows: [LimitWindow(kind: .fiveHour, percent: 1, resetsAt: Self.now)]))

        #expect(json.contains(##""2026-09-12T12:00:00\u002B00:00""##))
    }
}
